### ISSUE TITLE
Object safety violation reported with arbitrary self types in trait method

### ISSUE THREAD

**Issue by alexkirsz** [rust-lang/rust#110491]
I tried this code [(playground)](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=6642da297fbae52bd9569032588f5eaa):

```rust
#![feature(arbitrary_self_types)]

struct Arbitrary<T> where T: ?Sized {
    _t: core::marker::PhantomData<T>
}

impl<T> core::ops::Deref for Arbitrary<T> where T: ?Sized {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        todo!()
    }
}

trait BoxTrait {
    fn method(self: Box<Self>) {}
}

trait ArbitraryTrait {
    fn method(self: Arbitrary<Self>) {}
}

fn main() {
    let ptr: Box<dyn BoxTrait> = todo!();
    ptr.method();

    let ptr: Arbitrary<dyn ArbitraryTrait> = todo!();
    ptr.method();
}
```

I expected this to compile just fine: `ArbitraryTrait` should be object safe since its only method accepts an `Arbitrary` self object. In fact, this works for `Box` but not `Arbitrary`.

Instead, the following errors are reported:

```
error[[E0038]](https://doc.rust-lang.org/nightly/error_codes/E0038.html): the trait `ArbitraryTrait` cannot be made into an object
  --> src/main.rs:27:14
   |
20 |     fn method(self: Arbitrary<Self>) {}
   |                     --------------- help: consider changing method `method`'s `self` parameter to be `&self`: `&Self`
...
27 |     let ptr: Arbitrary<dyn ArbitraryTrait> = todo!();
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `ArbitraryTrait` cannot be made into an object
   |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
  --> src/main.rs:20:21
   |
19 | trait ArbitraryTrait {
   |       -------------- this trait cannot be made into an object...
20 |     fn method(self: Arbitrary<Self>) {}
   |                     ^^^^^^^^^^^^^^^ ...because method `method`'s `self` parameter cannot be dispatched on

error[[E0038]](https://doc.rust-lang.org/nightly/error_codes/E0038.html): the trait `ArbitraryTrait` cannot be made into an object
  --> src/main.rs:28:5
   |
20 |     fn method(self: Arbitrary<Self>) {}
   |                     --------------- help: consider changing method `method`'s `self` parameter to be `&self`: `&Self`
...
28 |     ptr.method();
   |     ^^^^^^^^^^^^ `ArbitraryTrait` cannot be made into an object
   |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
  --> src/main.rs:20:21
   |
19 | trait ArbitraryTrait {
   |       -------------- this trait cannot be made into an object...
20 |     fn method(self: Arbitrary<Self>) {}
   |                     ^^^^^^^^^^^^^^^ ...because method `method`'s `self` parameter cannot be dispatched on

For more information about this error, try `rustc --explain E0038`.
```

### Meta

`rustc --version --verbose`:
```
rustc 1.71.0-nightly (7908a1d65 2023-04-17)
binary: rustc
commit-hash: 7908a1d65496b88626e4b7c193c81d777005d6f3
commit-date: 2023-04-17
host: aarch64-apple-darwin
release: 1.71.0-nightly
LLVM version: 16.0.2
```

**Comment by SkiFire13** [comment]
I feel like this is correct, `Arbitrary<dyn ArbitraryTrait>` cannot possibly store the vtable for `ArbitraryTrait` and thus `ArbitraryTrait` can't be object safe. `Box` instead works because it can store the vtable by becoming a fat pointer. There could be a better error though.

**Comment by alexkirsz** [comment]
Could the compiler automatically exclude this method from the vtable, like it does for methods with a `Self: Sized` bound?

For my use case, I am working around this issue by using an `Arbitrary<Box<dyn ArbitraryTrait>>` instead and adding a `Self: Sized` bound to the trait method's declaration, but it feels like an arbitrary restriction. [Link to playground.](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=05b5aaa38fd882818d546cf2a08bcd18)

**Comment by SkiFire13** [comment]
If the compiler excludes it from the vtable then you won't be able to call it given an `Arbitrary<dyn ArbitraryTrait>`, i.e. when `Self = dyn ArbitraryTrait`, essentially ruling out `Self: !Sized`.

Why can't you add a `Self: Sized` bound to it?

**Comment by alexkirsz** [comment]
If I add a `Self: Sized` bound to the method, I can no longer call it directly on an `Arbitrary<dyn ArbitraryTrait>`, as `dyn ArbitraryTrait` is `!Sized`. Hence the need to use `Arbitrary<Box<dyn ArbitraryTrait>>` or `Arbitrary<&'static dyn ArbitraryTrait>` instead.

**Comment by SkiFire13** [comment]
> If I add a `Self: Sized` bound to the method, I can no longer call it directly on an `Arbitrary<dyn ArbitraryTrait>`

But that's also true if the method is removed from the vtable...

Let's try this: which implementation of `method` do you expect to be executed with this code?

```rust
let arbitrary = Arbitrary {
    _t: core::marker::PhantomData::<dyn ArbitraryTrait>
};

arbitrary.method();
```

**Comment by alexkirsz** [comment]
Right! I understand the issue now. Thanks for taking the time to walk me through it 😄
